casper/src/main/resources/RevVault.rho

```diff
@@ -27,255 +28,242 @@ in {
+              for (retCh <- createVault) {
+                _makeVault!(ownerRevAddress, 0, *retCh) |
+                for (_ <- ackCh) { Nil }
               }
-            } |
-            // Used during genesis to create PoS vault with initial balance = sum of initial bonds.
-            for (@"createWithBalance", @ownerRevAddress, @amount, ret <- RevVault) {
-              new createVault in {
-                _getOrCreate!(ownerRevAddress, *createVault, *ret) |
-                for (retCh <- createVault) {
-                  _makeVault!(ownerRevAddress, amount, *retCh)
-                }
+            }
+          } |
+          // Used during genesis to create PoS vault with initial balance = sum of initial bonds.
+          for (@"createWithBalance", @ownerRevAddress, @amount, ret <- RevVault) {
+            new createVault, ackCh in {
+              _getOrCreate!(ownerRevAddress, *createVault, *ret, *ackCh) |
+              for (retCh <- createVault) {
+                _makeVault!(ownerRevAddress, amount, *retCh) |
+                for (_ <- ackCh) { Nil }
               }
-            } |
-
-            contract _getOrCreate(@revAddress, constructor, retCh) = {
-              new revAddressValidCh,
-                  revAddressValidEitherCh,
-                  getOrCreate
-              in {
-                RevAddress!("validate", revAddress, *revAddressValidCh) |
-                @Either!("fromNillableError <-", *revAddressValidCh, *revAddressValidEitherCh) |
-                @Either!("flatMap <-", *revAddressValidEitherCh, *getOrCreate, *retCh) |
-                for (_, ret <- getOrCreate) {
-                  // At first peek for `vaultMap` to check if vault (address) exists
-                  // - peek ensures that for existing vaults we don't have a block conflict
-                  for (@vaultMap <<- vaultMapStore) {
-                    match vaultMap.get(revAddress) {
-                      Nil => {
-                        // Vault is not found, create new and save it in the `vaultMap`
-                        // - this means two blocks with new vaults cannot be merged
-                        for (@vaultMap <- vaultMapStore) {
-                          // We need to check again because `vaultMap` is not locked with peek
-                          match vaultMap.get(revAddress) {
-                            Nil => {
-                              new resCh in {
-                                constructor!(*resCh) |
-                                for (@eitherVault <- resCh) {
-                                  ret!(eitherVault) |
-                                  match eitherVault {
-                                    (true, vault) => vaultMapStore!(vaultMap.set(revAddress, vault))
-                                    (false, _)    => vaultMapStore!(vaultMap)
-                                  }
+            }
+          } |
+
+          contract _getOrCreate(@revAddress, constructor, retCh, ackCh) = {
```

Comment:
Why is needed this additional _ackCh_ when there is already _retCh_?

Reply:
You're right! It's better to create a new _ackCh_ inside of __getOrCreate_ and consume the acknowledgement there. Thanks!